lib/compat/wordpress-6.8/class-gutenberg-hierarchical-sort.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+/**
+ * A custom REST server for Gutenberg.
+ *
+ * @package gutenberg
+ * @since   6.8.0
+ */
+
+class Gutenberg_Hierarchical_Sort {
+	private static $post_ids = array();
+	private static $levels   = array();
+	private static $instance;
+
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	public function run( $args ) {
+		$new_args = array_merge(
+			$args,
+			array(
+				'fields'         => 'id=>parent',
+				'posts_per_page' => -1,
+			)
+		);
+		$query    = new WP_Query( $new_args );
+		$posts    = $query->posts;
+		$result   = self::sort( $posts );
+
+		self::$post_ids = $result['post_ids'];
+		self::$levels   = $result['levels'];
+	}
+
+	/**
+	 * Check if the request is eligible for hierarchical sorting.
+	 *
+	 * @param array $request The request data.
+	 *
+	 * @return bool Return true if the request is eligible for hierarchical sorting.
+	 */
+	public static function is_eligible( $request ) {
+		if ( ! isset( $request['orderby_hierarchy'] ) || true !== $request['orderby_hierarchy'] ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	public static function get_ancestor( $post_id ) {
+		return get_post( $post_id )->post_parent ?? 0;
+	}
+
+	/**
+	 * Sort the post ids by hierarchy.
+	 *
+	 * Example: $posts is
+	 * [
+	 *   ['ID' => 4, 'post_parent' => 2],
+	 *   ['ID' => 2, 'post_parent' => 0],
+	 *   ['ID' => 3, 'post_parent' => 2],
+	 * ]
+	 *
+	 * and we want to return: [2, 3, 4]
+	 *
+	 * @param array $posts The posts to sort.
+	 * @param array $args  The arguments to sort the posts by.
+	 *
+	 * @return array Return the sorted post_ids and the corresponding levels.
+	 */
+	public static function sort( $posts ) {
+		/*
+		 * Arrange pages in two arrays:
+		 *
+		 * - $top_level: posts whose parent is 0
+		 * - $children: post ID as the key and an array of children post IDs as the value.
+		 *   Example: $children[10][] contains all sub-pages whose parent is 10.
+		 *
+		 * Additionally, keep track of the levels of each post in $levels.
+		 * Example: $levels[10] = 0 means the post ID is a top-level page.
+		 *
+		 */
+		$top_level = array();
+		$children  = array();
+		foreach ( $posts as $post ) {
+			if ( 0 === $post->post_parent ) {
+				$top_level[] = $post->ID;
+			} else {
+				$children[ $post->post_parent ][] = $post->ID;
```

Comment:
TIL I learned that PHP arrays are implicitly created when attempting to push to an undefined variable. Works for atomic identifiers too:

```sh
php -r '$a[] = 42; print_r($a);'
```
```
Array
(
    [0] => 42
)
```

Reply:
I pushed https://github.com/WordPress/gutenberg/pull/66479/commits/307944daf9a06ff36fc5d55c5e973a8c0de4d08e to make sure the key is never undefined or null.

Though, to be honest, I don't know that it can be ever be [in that state](https://github.com/WordPress/wordpress-develop/blob/ffceac51dab7878c1401c2a1cc7737409fcff7a9/src/wp-includes/class-wp-post.php#L168). But we do cover against emptiness in [other places](https://github.com/WordPress/wordpress-develop/blob/ffceac51dab7878c1401c2a1cc7737409fcff7a9/src/wp-includes/post-template.php#L2032), so being cautious doesn't hurt.

Reply:
No no, I was referring to something else, and there was no action needed here. :)

I saw this line:

```js
$children[ $post->post_parent ][] = $post->ID;
```

and noticed that the subarrays are never created, i.e. there is nothing in this diff like:

```js
if ( ! isset( $children[ $post->post_parent ] ) ) {
  $children[ $post->post_parent ] = array();
}
```

which prompted me to test and realise that PHP doesn't care: if it sees a push operation (`[]`) is implicitly creates an empty array there. So `$children[ $foo ][] = 'bla bla'` will ensure there is an array containing at least `'bla bla'` at `$children[ $foo ]`.